concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+        put qcEpoch
+        put qcAggregateSignature
+        put qcSignatories
+        put feSuccessorProof
+    get = do
+        feFinalizedQuorumCertificate <- get
+        qcEpoch <- get
+        qcAggregateSignature <- get
+        qcSignatories <- get
+        feSuccessorProof <- get
+        let sqcRound = qcRound feFinalizedQuorumCertificate + 1
+        let feSuccessorQuorumCertificate =
+                QuorumCertificate
+                    { qcRound = sqcRound,
+                      qcBlock =
+                        successorBlockHash
+                            sqcRound
+                            qcEpoch
+                            (qcBlock feFinalizedQuorumCertificate)
+                            feSuccessorProof,
+                      ..
+                    }
+        return FinalizationEntry{..}
+
+instance HashableTo Hash.Hash (Optionally FinalizationEntry) where
+    getHash Absent = Hash.hash $ encode (0 :: Word8)
+    getHash (Present fe) = Hash.hash $ runPut $ do
+        putWord8 1
+        put fe
+
+data TimeoutSignatureMessage = TimeoutSignatureMessage
+    { -- |Hash of the genesisBlock
+      tsmGenesis :: !BlockHash,
+      -- |Round number of the timed-out round
+      tsmRound :: !Round,
+      -- |Round number of the highest known valid quorum certificate
+      tsmQCRound :: !Round
+    }
+    deriving (Eq, Show)
+
+-- |Compute the byte representation of a 'TimeoutSignatureMessage' that is actually signed.
+-- TODO: check that this cannot collide with other signed messages.
+timeoutSignatureMessageBytes :: TimeoutSignatureMessage -> BS.ByteString
+timeoutSignatureMessageBytes TimeoutSignatureMessage{..} = runPut $ do
+    putByteString "TIMEOUT."
+    put tsmGenesis
+    put tsmRound
+    put tsmQCRound
+
+-- |Signature by a finalizer on a 'TimeoutSignatureMessage', or an aggregation of signatures on a
+-- common message.
+newtype TimeoutSignature = TimeoutSignature {theTimeoutSignature :: Bls.Signature}
+    deriving (Eq, Ord, Show, Serialize, Semigroup, Monoid)
+
+-- |Sign a 'TimeoutSignatureMessage' with a Baker's private key.
+signTimeoutSignatureMessage :: TimeoutSignatureMessage -> BakerAggregationPrivateKey -> TimeoutSignature
+signTimeoutSignatureMessage msg privKey =
+    TimeoutSignature $ Bls.sign (timeoutSignatureMessageBytes msg) privKey
+
+-- |Check the signature on a 'TimeoutSignatureMessage' that is signed by a single party.
+checkTimeoutSignatureSingle ::
+    TimeoutSignatureMessage -> BakerAggregationVerifyKey -> TimeoutSignature -> Bool
+checkTimeoutSignatureSingle msg pubKey =
+    Bls.verify (timeoutSignatureMessageBytes msg) pubKey
+        . theTimeoutSignature
+
+-- |Data structure recording which finalizers have quorum certificates for which rounds.
+--
+-- Invariant: @Map.size theFinalizerRounds <= fromIntegral (maxBound :: Word32)@.
+-- (This is trivially satisfied if each finalizer index occurs for at most one round.)
```

Comment:
Isn't this technically off by one? There are 2^32 possible finalizer indices, so theoretically maxBound + 1 is the size of the set even if this property is satisfied.

Reply:
Removed the comment in the parentheses.